### ISSUE TITLE
Rstudio freezing up immediately after prediction

### ISSUE THREAD

**Issue by quantitative-technologies** [rstudio/keras#957]
I'm not quite sure where to leave this report, but it is a very serious issue (bad enough that I am seriously considering switching the project to Python), and I've only seen it with R Keras.

It has happened on two different Linux installations, but the same code does not have this issue on Windows.

When calling the line of code:

    evaluation <- Model %>% evaluate(X, Y)

with a `keras` model `Model` Rstudio freezes up for several minutes (last time I timed it as 8:30 min) with the process `QtWebEngineProc` running at 100%. 

I have confirmed by wrapping the above code with a `system.time` call that the above call itself is not getting stuck. However, the output in the R console is abnormal. The equal signs nearly fill the whole R console, and when I tried to copy and paste it here it exceeds the github size limit:

Epoch 1/3
272115/272115 [==============================] - 6s 21us/sample - loss: 0.7082 - accuracy: 0.5320
Epoch 2/3
272115/272115 [==============================] - 8s 31us/sample - loss: 0.6906 - accuracy: 0.5329
Epoch 3/3
272115/272115 [==============================] - 5s 19us/sample - loss: 0.6884 - accuracy: 0.5336
90782/1 [==========================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================
... 
cut out "thousands" of lines of the same
...
==================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================] - 1s 14us/sample - loss: 0.6967 - accuracy: 0.5182
Evaluated model on test set: 1.345 sec elapsed
2.025 0.148 1.346 0 0Called from: neural.predict.keras(1, modeling_data$test, feature_params$feature_cols, 
    label_col, label_type = "binary")

> sessionInfo()
R version 3.6.1 (2019-07-05)
Platform: x86_64-pc-linux-gnu (64-bit)
Running under: Ubuntu 18.04.2 LTS

Matrix products: default
BLAS:   /usr/lib/x86_64-linux-gnu/openblas/libblas.so.3
LAPACK: /usr/lib/x86_64-linux-gnu/libopenblasp-r0.2.20.so

locale:
 [1] LC_CTYPE=en_CA.UTF-8       LC_NUMERIC=C               LC_TIME=en_CA.UTF-8        LC_COLLATE=en_CA.UTF-8    
 [5] LC_MONETARY=en_CA.UTF-8    LC_MESSAGES=en_CA.UTF-8    LC_PAPER=en_CA.UTF-8       LC_NAME=C                 
 [9] LC_ADDRESS=C               LC_TELEPHONE=C             LC_MEASUREMENT=en_CA.UTF-8 LC_IDENTIFICATION=C       

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
 [1] here_0.1          tictoc_1.0        purrr_0.3.3       data.table_1.12.6 caret_6.0-84      ggplot2_3.2.1    
 [7] lattice_0.20-38   keras_2.2.5.0     rlang_0.4.2       magrittr_1.5      assertthat_0.2.1  digest_0.6.22    

loaded via a namespace (and not attached):
 [1] reticulate_1.14    tidyselect_0.2.5   reshape2_1.4.3     splines_3.6.1      colorspace_1.4-1   generics_0.0.2    
 [7] stats4_3.6.1       base64enc_0.1-3    survival_2.44-1.1  prodlim_2019.11.13 ModelMetrics_1.2.2 pillar_1.4.2      
[13] glue_1.3.1         withr_2.1.2        rappdirs_0.3.1     foreach_1.4.7      plyr_1.8.4         tensorflow_2.0.0  
[19] lava_1.6.6         stringr_1.4.0      timeDate_3043.102  munsell_0.5.0      gtable_0.3.0       recipes_0.1.7     
[25] codetools_0.2-16   tfruns_1.4         class_7.3-15       Rcpp_1.0.3         backports_1.1.5    scales_1.0.0      
[31] ipred_0.9-9        jsonlite_1.6       stringi_1.4.3      dplyr_0.8.3        rprojroot_1.3-2    grid_3.6.1        
[37] tools_3.6.1        lazyeval_0.2.2     tibble_2.1.3       crayon_1.3.4       whisker_0.4        pkgconfig_2.0.3   
[43] zeallot_0.1.0      MASS_7.3-51.4      Matrix_1.2-17      lubridate_1.7.4    gower_0.2.1        rstudioapi_0.10   
[49] iterators_1.0.12   R6_2.4.1           rpart_4.1-15       nnet_7.3-12        nlme_3.1-141       compiler_3.6.1

**Comment by quantitative-technologies** [comment]
Update: After submitting this report, I noticed a freeze up but this time during training.

**Comment by skeydan** [comment]
Can you provide a minimal reproducible example?

**Comment by quantitative-technologies** [comment]
I can't spend too much time on this right now, trying to minimize the example, but let me see what I can come up with...

**Comment by quantitative-technologies** [comment]
First I am looking into problems at the OS level. It looks like there may be GPU memory issues. Does this look normal, Xorg and gnome-shell using 5GB of video RAM?

    $ nvidia-smi 
    Thu Jan  2 06:58:35 2020       
    +-----------------------------------------------------------------------------+
    | NVIDIA-SMI 418.56       Driver Version: 418.56       CUDA Version: 10.1     |
    |-------------------------------+----------------------+----------------------+
    | GPU  Name        Persistence-M| Bus-Id        Disp.A | Volatile Uncorr. ECC |
    | Fan  Temp  Perf  Pwr:Usage/Cap|         Memory-Usage | GPU-Util  Compute M. |
    |===============================+======================+======================|
    |   0  GeForce GTX 1080    Off  | 00000000:01:00.0  On |                  N/A |
    | N/A   54C    P2    38W /  N/A |   7908MiB /  8085MiB |      0%      Default |
    +-------------------------------+----------------------+----------------------+
                                                                                   
    +-----------------------------------------------------------------------------+
    | Processes:                                                       GPU Memory |
    |  GPU       PID   Type   Process name                             Usage      |
    |=============================================================================|
    |    0       342      G   krusader                                      34MiB |
    |    0      3202      G   /usr/lib/xorg/Xorg                            40MiB |
    |    0      3298      G   /usr/bin/gnome-shell                          16MiB |
    |    0      7278      G   /snap/pycharm-community/172/jbr/bin/java      35MiB |
    |    0     16731      G   /usr/lib/xorg/Xorg                          2599MiB |
    |    0     16903      G   /usr/bin/gnome-shell                        2298MiB |
    |    0     17401      G   ...-token=4448E31B404BA6C80C5468520EB15F06   177MiB |
    |    0     20967      G   /usr/lib/rstudio/bin/rstudio                 228MiB |
    |    0     21136      C   /usr/lib/rstudio/bin/rsession               2455MiB |
    |    0     22856      G   krusader                                       2MiB |
    |    0     25788      G   gnome-control-center                           2MiB |
    |    0     30720      G   /opt/teamviewer/tv_bin/TeamViewer              9MiB |
    +-----------------------------------------------------------------------------+

**Comment by skeydan** [comment]
This looks relevant: https://bugs.launchpad.net/ubuntu/+source/gnome-shell/+bug/1823544

(I haven't seen such a situation myself.)

How long has the machine been running? Try rebooting if possible.

**Comment by quantitative-technologies** [comment]
Thanks, I will look into that if the problem persists.

23 days. One of the reasons I like Linux. 

I just upgraded from Ubuntu 18.04 to 19.10, and will reboot now.

**Comment by quantitative-technologies** [comment]
I am still looking in to the gnome-shell and Xorg memory usage which persisted after the upgrade.

However, even after reboot the issue with RStudio freezing remained. Also, I moved the ML stuff from this project to Python and there are no issues there. 

I am too busy to work on coming up with an MRE, but I could send a serialized file with the model and data in case that is helpful.

**Comment by skeydan** [comment]
I think that in your case - given you still have the gnome-shell issues, so there's inevitably a shortage of GPU memory - I'd try running the model from R instead of RStudio. Does that work?

**Comment by fullmetalomelette** [comment]
I am facing the same issue that the OP described. Running model evaluate() within RStudio results in `QtWebEngineProc` going to 100% and the system freezing for a solid ~5 minutes before finally returning the command, as well as printing equal signs that fill up the entire console. When I run the exact same command in R on the cmd-line there is no freezing/lag and the evaluation executes immediately with no printing of equal signs.

Reproducible example:
`max_features <- 10000                                         
maxlen <- 20                                                  
imdb <- dataset_imdb(num_words = max_features)
c(c(x_train, y_train), c(x_test, y_test)) %<-% imdb           
x_train <- pad_sequences(x_train, maxlen = maxlen) # Restrict each entry to the first 20 words    
x_test <- pad_sequences(x_test, maxlen = maxlen)

model <- keras_model_sequential() %>%
  layer_embedding(input_dim = 10000, output_dim = 8,            
                  input_length = maxlen) %>%
  layer_flatten() %>%                                           
  layer_dense(units = 1, activation = "sigmoid")                
model %>% compile(
  optimizer = "rmsprop",
  loss = "binary_crossentropy",
  metrics = c("acc")
)
summary(model)
history <- model %>% fit(
  x_train, y_train,
  epochs = 10,
  batch_size = 32,
  validation_split = 0.2
)

model %>% evaluate(x_test, y_test)`

`> sessionInfo()
R version 3.6.0 (2019-04-26)
Platform: x86_64-apple-darwin15.6.0 (64-bit)
Running under: macOS Mojave 10.14.6

Matrix products: default
BLAS:   /System/Library/Frameworks/Accelerate.framework/Versions/A/Frameworks/vecLib.framework/Versions/A/libBLAS.dylib
LAPACK: /Library/Frameworks/R.framework/Versions/3.6/Resources/lib/libRlapack.dylib

locale:
[1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
 [1] forcats_0.4.0   stringr_1.4.0   dplyr_0.8.5     purrr_0.3.3     readr_1.3.1    
 [6] tidyr_1.0.0     tibble_2.1.3    ggplot2_3.3.0   tidyverse_1.2.1 keras_2.2.5.0  

loaded via a namespace (and not attached):
 [1] reticulate_1.14    tidyselect_1.0.0   haven_2.2.0        lattice_0.20-38   
 [5] colorspace_1.4-1   vctrs_0.2.4        generics_0.0.2     getopt_1.20.3     
 [9] base64enc_0.1-3    blob_1.1.1         rlang_0.4.5        pillar_1.4.2      
[13] withr_2.1.2        optparse_1.6.2     glue_1.3.1         DBI_1.0.0         
[17] tidydataclient_1.0 bit64_0.9-7        modelr_0.1.4       readxl_1.3.1      
[21] lifecycle_0.1.0    tensorflow_2.0.0   munsell_0.5.0      gtable_0.3.0      
[25] cellranger_1.1.0   rvest_0.3.4        memoise_1.1.0      tfruns_1.4        
[29] broom_0.5.2        logging_0.9-107    Rcpp_1.0.1         backports_1.1.4   
[33] scales_1.1.0       jsonlite_1.6       bit_1.1-14         hms_0.5.3         
[37] packrat_0.5.0      digest_0.6.24      stringi_1.4.3      grid_3.6.0        
[41] cli_1.1.0          fasttime_1.0-2     tools_3.6.0        magrittr_1.5      
[45] RSQLite_2.1.1      tidydatar_1.0      crayon_1.3.4       whisker_0.3-2     
[49] pkgconfig_2.0.2    zeallot_0.1.0      xml2_1.2.0         lubridate_1.7.4   
[53] assertthat_0.2.1   httr_1.4.1         rstudioapi_0.10    R6_2.4.1          
[57] nlme_3.1-140       compiler_3.6.0`
